### PR TITLE
[core] Deduplicate repetitive logs across the cluster

### LINKED ISSUE #33184: [Ray Core] Deduplicate logging messages from log_to_driver
### Description

Currently if a Ray cluster has many workers, it can happen that all workers log the same message through the `log_to_driver` mechanism. This can lead to a lot of log spam which is annoying to handle. It would be much better to dedup those messages.

### Use case

_No response_

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

User code and libraries often emit repetitive log messages, which is a pain point for large-scale cluster runs. Add a mechanism to automatically group together and deduplicate these messages.

Sample code:
```
import ray
import random

@ray.remote
def tf():
    import tensorflow
    print("Hello there, I just imported tensorflow!", random.random())

ray.get([tf.remote() for _ in range(100)])
```

Sample output:
```
(tf pid=319406) 2023-03-16 17:22:12.796927: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F AVX512_VNNI FMA
(tf pid=319406) To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
(tf pid=319406) 2023-03-16 17:22:12.930249: I tensorflow/core/util/port.cc:104] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
(tf pid=319406) 2023-03-16 17:22:12.934243: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
(tf pid=319406) 2023-03-16 17:22:12.934263: I tensorflow/compiler/xla/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
(tf pid=319399) 2023-03-16 17:22:13.890206: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.
(tf pid=319399) /home/eric/.local/lib/python3.8/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.1
(tf pid=319399)   warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"
(tf pid=319399) Hello there, I just imported tensorflow! 0.031252488755938335
(tf pid=319406) Hello there, I just imported tensorflow! 0.5185359453242194 [repeated 99x across cluster] (set RAY_DEDUP_LOGS=0 to disable log deduplication)
(tf pid=319400) 2023-03-16 17:22:12.896545: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F AVX512_VNNI FMA [repeated 7x across cluster]
(tf pid=319400) To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags. [repeated 7x across cluster]
(tf pid=319400) 2023-03-16 17:22:13.059681: I tensorflow/core/util/port.cc:104] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`. [repeated 7x across cluster]
(tf pid=319401) 2023-03-16 17:22:14.036612: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libnvinfer_plugin.so.7: cannot open shared object file: No such file or directory [repeated 23x across cluster]
(tf pid=319400) 2023-03-16 17:22:13.063986: I tensorflow/compiler/xla/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine. [repeated 7x across cluster]
(tf pid=319401) 2023-03-16 17:22:14.036620: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly. [repeated 7x across cluster]
(tf pid=319401) /home/eric/.local/lib/python3.8/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.1 [repeated 7x across cluster]
(tf pid=319401)   warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}" [repeated 7x across cluster]

```

Credit: @hauntsaninja for original implementation
Closes: https://github.com/ray-project/ray/issues/33184
## Checks

TODO:
- [x] Add unit tests
- [x] Add output tests

### CODE REVIEW

**Review by rkooo567 — COMMENTED** [review]
@ericl should I take over this PR? I think we need additional work;

1. add tests 
2. Doc changes
3. several test failures

Also, should we turn off by default first? I think it could be risky to turn this on in the next version given the branch cut is very soon, and we don't have enough time testing?

File: python/ray/_private/ray_logging.py [hunk 1]
```diff
@@ -253,3 +261,153 @@ def emit(self, data):
...
+        """Rewrite a batch of lines to reduce duplicate log messages.
+
+        Args:
+            lines: The batch of lines from a single source.
+            metadata: Dict describing the data source.
+
+        Returns:
+            Batch of deduplicated lines.
+        """
+        global _log_dedup_warned
```
Comment (rkooo567) [inline python/ray/_private/ray_logging.py:313]: This variable doesn't seem to be used

File: python/ray/_private/ray_logging.py [hunk 2]
```diff
@@ -253,3 +261,153 @@ def emit(self, data):
...
+                    )
+                )
+            elif state.count > 0:
+                output.append(dict(state.metadata, **{"lines": [state.line]}))
+        self.recent.clear()
+        return output
+
+    def _warn_once(self) -> str:
+        if log_once("log_dedup_warning"):
+            return " (set RAY_DEDUP_LOGS=0 to disable log deduplication)"
```
Comment (rkooo567) [inline python/ray/_private/ray_logging.py:404]:
```suggestion
            return " (Ray deduplicates logs by default. set RAY_DEDUP_LOGS=0 to disable log deduplication)"
```
?

File: python/ray/_private/ray_logging.py [hunk 3]
```diff
@@ -253,3 +261,153 @@ def emit(self, data):
...
+                else:
+                    # Don't dedup messages from the same source, just print.
+                    output.append(line)
+            else:
+                self.recent[dedup_key] = DedupState(now, 0, line, metadata, {source})
+                output.append(line)
+
+        # Flush patterns from the buffer that are older than the aggregation window.
+        while self.recent:
+            if now - next(iter(self.recent.values())).timestamp > AGGREGATION_WINDOW_S:
```
Comment (rkooo567) [inline python/ray/_private/ray_logging.py:348]: Looks like the if here is False, `dedup_key = next(iter(self.recent))` can be buggy?

Hmm I found if it is else, we just break. Maybe we can refactor code this way to make this part clearer? 

```
dedup_state = next(iter(self.recent.values())
if now - dedup.state.timestamp < AGGREGATION_WINDOW_S:
    break

dedup_key = next(iter(self.recent))
...
```

File: python/ray/_private/ray_logging.py [hunk 4]
```diff
@@ -253,3 +261,153 @@ def emit(self, data):
...
+            List of log batches to print.
+        """
+        output = []
+        for state in self.recent.values():
+            if state.count > 1:
+                output.append(
+                    dict(
+                        state.metadata,
+                        **{
+                            "lines": [
```
Comment (rkooo567) [inline python/ray/_private/ray_logging.py:387]: This logic seems to be the same as one inside `deduplicate`. Should we take it out to a helper method?

File: python/ray/_private/ray_logging.py [hunk 5]
```diff
@@ -253,3 +261,153 @@ def emit(self, data):
...
+                    self.recent[dedup_key] = state
+                elif state.count > 0:
+                    # Aggregation wasn't fruitful, print the line and stop aggregating.
+                    output.append(line)
+            else:
+                break
+
+        return output
+
+    def flush(self) -> List[dict]:
```
Comment (rkooo567) [inline python/ray/_private/ray_logging.py:374]: Actually if we make `deduplicate` to return the same output (metadata and lines pair), we can just reuse implementation of `deduplicate` from `flush`?

Isn't this method same as the the `duplicate` with AGGREGATION_WINDOW_S 0 seconds?

**Comment by ericl** [comment]
Yeah, this isn't a priority for 2.4. Actually, could you try this out on some different libraries? I think we need to do some more testing before we merge, in case the algorithm should be tweaked (i.e., mangles some library outputs by default).

**Comment by pcmoritz** [comment]
Tune test case:

```python
import time
import warnings

import ray
from ray.air import session
from ray import tune


def evaluation_fn(step, width, height):
    time.sleep(0.1)
    return (0.1 + width * step / 100) ** (-1) + height * 0.1


def easy_objective(config):
    # Hyperparameters
    width, height = config["width"], config["height"]

    warnings.warn("hello, world")

    for step in range(config["steps"]):
        # Iterative training function - can be any arbitrary training procedure
        intermediate_score = evaluation_fn(step, width, height)
        # Feed the score back back to Tune.
        session.report({"iterations": step, "mean_loss": intermediate_score})


# This will do a grid search over the `activation` parameter. This means
# that each of the two values (`relu` and `tanh`) will be sampled once
# for each sample (`num_samples`). We end up with 2 * 50 = 100 samples.
# The `width` and `height` parameters are sampled randomly.
# `steps` is a constant parameter.

tuner = tune.Tuner(
        easy_objective,
        tune_config=tune.TuneConfig(
            metric="mean_loss",
            mode="min",
            num_samples=5,
        ),
        param_space={
            "steps": 5,
            "width": tune.uniform(0, 20),
            "height": tune.uniform(-100, 100),
            "activation": tune.grid_search(["relu", "tanh"]),
        },
)
results = tuner.fit()

print("Best hyperparameters found were: ", results.get_best_result().config)
```

**Comment by pcmoritz** [comment]
I played around with this a bunch, also with some tasks etc. and sometimes the logs show up a little later than expected, that's the only downside I have seen so far. Why don't we get it merged and we can iterate on it a little more.

The code looks great to me! Thanks for doing this!

**Review by pcmoritz — APPROVED** [review]

**Comment by ericl** [comment]
Sounds good. Let me add a few basic tests tomorrow and we can try it out.

**Comment by cadedaniel** [comment]
This is unhelpful for SPMD development use-cases, where advanced users use actors to each run the same script. The relative ordering of logs wrt to the other logs from the same process is important signal when debugging hangs and nondeterministic behavior.

**Comment by ericl** [comment]
Makes sense. Does it work to set RAY_DEDUP_LOGS=0 for these use cases?

**Comment by cadedaniel** [comment]
Yes, that solves the issue.

Bigger picture, this feels like we're trading off Ray Core versatility for ease-of-use for simple use cases. Can we push this functionality up the stack so that the libraries opt-in to enable it?

btw great feature, for many cases this really helpful :)

**Comment by ericl** [comment]
Hmm, I don't think there is any tradeoff here. You can always disable it, and expect maybe most users will need to disable it at one point or another.

**Comment by cadedaniel** [comment]
> Hmm, I don't think there is any tradeoff here.

The tradeoff is clear to me. Users looking for "distributed Python" are not all looking for "distributed Python + opinionated log aggregation". Putting "opinionated log aggregation" on the default path means friction for those who don't care for it, hence trading off usability for those who don't for those who do.

I can't emphasize this enough, it is a *massive* pain to reproduce and understand certain classes of hangs or inconsistent incorrectness issues when training LLMs. Things like race conditions between when different CPU threads interact with GPU streams, that only occur at larger scale (ask me about a bug in activation offloading when training GPT-2 175B that only reproduced on >=64 p4d's).

Tools like OpenMPI (which otherwise are very limited compared to Ray) get this right by taking an unopinionated stance on log contents.

**Comment by ericl** [comment]
@pcmoritz any thoughts?

**Pull request merged**